lang/csharp/src/apache/main/IO/Parsing/Symbol.cs

```diff
@@ -0,0 +1,983 @@
+            public byte[] Contents { get; private set; }
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="Symbol.DefaultStartAction"/> class.
+            /// </summary>
+            public DefaultStartAction(byte[] contents)
+            {
+                Contents = contents;
+            }
+        }
+
+        /// <summary>
+        /// The union adjust action.
+        /// </summary>
+        public class UnionAdjustAction : ImplicitAction
+        {
+            /// <summary>
+            /// The index.
+            /// </summary>
+            public int RIndex { get; private set; }
+
+            /// <summary>
+            /// The symbol to parser.
+            /// </summary>
+            public Symbol SymToParse { get; private set; }
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="Symbol.UnionAdjustAction"/> class.
+            /// </summary>
+            public UnionAdjustAction(int rindex, Symbol symToParse)
+            {
+                RIndex = rindex;
+                SymToParse = symToParse;
+            }
+
+            /// <inheritdoc />
+            protected override Symbol Flatten(IDictionary<Sequence, Sequence> map,
+                IDictionary<Sequence, IList<Fixup>> map2)
+            {
+                return new UnionAdjustAction(RIndex, SymToParse.Flatten(map, map2));
+            }
+        }
+
+        /// <summary>
+        /// The enum labels action.
+        /// </summary>
+        public class EnumLabelsAction : IntCheckAction
+        {
+            /// <summary>
+            /// The symbols.
+            /// </summary>
+            public IList<string> Symbols { get; private set; }
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="Symbol.EnumLabelsAction"/> class.
+            /// </summary>
+            public EnumLabelsAction(IList<string> symbols) : base(symbols.Count)
+            {
+                Symbols = symbols;
+            }
+
+            /// <summary>
+            /// Returns the label at the given index.
+            /// </summary>
+            public virtual string GetLabel(int n)
+            {
+                return Symbols[n];
+            }
+
+            /// <summary>
+            /// Returns index of the given label.
+            /// </summary>
+            public virtual int FindLabel(string l)
+            {
+                if (!ReferenceEquals(l, null))
+                {
+                    for (int i = 0; i < Symbols.Count; i++)
+                    {
+                        if (l.Equals(Symbols[i]))
+                        {
+                            return i;
+                        }
+                    }
+                }
+
+                return -1;
+            }
+        }
+
+        /// <summary>
+        /// The terminal symbols for the grammar.
+        /// </summary>
+        public static readonly Symbol Null = new Terminal("null");
```

Comment:
`public static Symbol Null => new Terminal("null");`
We should change these to static getters.  This is more of a common implementation in C# that is expected.

Reply:
Thanks, I went with `{ get; }` as the code does `==` checks on the symbols.